src/dma.c

```diff
@@ -47,42 +47,230 @@ static volatile struct PI_regs_s * const PI_regs = (struct PI_regs_s *)0xa460000
  *
  * @return nonzero if the DMA controller is busy or 0 otherwise
  */
-volatile int dma_busy() 
+volatile int __dma_busy(void)
 {
     return PI_regs->status & (PI_STATUS_DMA_BUSY | PI_STATUS_IO_BUSY);
 }
 
+volatile int dma_busy(void)
+{
+    return __dma_busy();
+}
+
 /**
- * @brief Read from a peripheral
+ * @brief Start reading data from a peripheral through PI DMA (low-level)
  *
- * This function should be used when reading from the cartridge.
+ * This function should be used when reading from a cartridge peripheral (typically
+ * ROM). This function just begins executing a raw DMA transfer, which is
+ * well-defined only for RAM addresses which are multiple of 8, ROM addresses
+ * which are  multiple of 2, and lengths which are multiple of 2.
+ * 
+ * Use #dma_wait to wait for the end of the transfer.
+ * 
+ * See #dma_read_async for a higher level primitive which can perform almost
+ * arbitrary transfers.
  *
  * @param[out] ram_address
- *             Pointer to a buffer to place read data
+ *             Pointer to a buffer to place read data (must be 8-byte aligned)
  * @param[in]  pi_address
- *             Memory address of the peripheral to read from
+ *             Memory address of the peripheral to read from (must be 2-byte aligned)
  * @param[in]  len
- *             Length in bytes to read into ram_address
+ *             Length in bytes to read into ram_address (must be multiple of 2)
  */
-void dma_read(void * ram_address, unsigned long pi_address, unsigned long len) 
+void dma_read_raw_async(void * ram_address, unsigned long pi_address, unsigned long len) 
 {
     assert(len > 0);
 
     disable_interrupts();
 
-    while (dma_busy()) ;
+    while (__dma_busy()) ;
     MEMORY_BARRIER();
     PI_regs->ram_address = ram_address;
     MEMORY_BARRIER();
     PI_regs->pi_address = (pi_address | 0x10000000) & 0x1FFFFFFF;
     MEMORY_BARRIER();
     PI_regs->write_length = len-1;
     MEMORY_BARRIER();
-    while (dma_busy()) ;
 
     enable_interrupts();
 }
 
+/** @brief Low-level 32-bit aligned ROM read.
+ * 
+ * @note This function must be called with interrupts disabled.
+ */
+static uint32_t __io_read32(unsigned long pi_address) {
+    while (__dma_busy()) {}
+    MEMORY_BARRIER();
+    return *(volatile uint32_t*)pi_address;
+}
+
+/** @brief Low-level 32-bit unaligned PI ROM read.
+ * 
+ * @note This function must be called with interrupts disabled.
+ */
+static uint32_t __io_read32u(unsigned long pi_address) {
+    uint32_t val = 0;
+
+    // We need to manually issue assembly opcodes here because we must
+    // wait for #dma_busy inbetween. So we can't simply tell GCC to 
+    // generate the LWL/LWR pair via __attribute__((aligned(1)), otherwise
+    // we would only be able to wait once before both of them.
+    while (__dma_busy()) {}
+    MEMORY_BARRIER();
+    __asm volatile("lwl %0,0(%1)" : "+r"(val) : "r"(pi_address));    
```

Comment:
Technically it does (both LWL and LWR only write a subset of bytes of the register). In this case, though, we know that LWR will change the bytes that LWL keeps intact. So we could indeed shortcut this and declare that LWL doesn't use the input value. But that would be false semantically speaking, and only true because there's a perfectly matching LWR later. So all in all, I'd rather keep the asm constraints semantically correct.